### PR TITLE
NICPS-19 : Search and Delete

### PR DESCRIPTION
**Pull request by chinmay15**
Problem : Provide support for search and delete items in all users.

Fix :
1) Existing perl script available at http://zre-matrix.eng.zimbra.com:8080/@md=d&cd=/depot/psteam/main/tools/&c=79S@/depot/psteam/main/tools/zmsearchNdelete/?ac=83 is added.
2) usage info and details are given in script itself

Testing done : Tested script with single and multi-node environment.

Testing needed by QA : Extensive testing is needed on single and multi-node environment with different type of search criteria.

Linked PR - https://github.com/Zimbra/zm-build/pull/79
https://github.com/Zimbra/zm-core-utils/pull/10

### CODE REVIEW

**Comment by chinmay15** [comment]
Snickers branch was deleted and re-created, so had to generate new pull request.
Old PR link - https://github.com/Zimbra/zm-core-utils/pull/10

File: src/libexec/zmsearchndelete [hunk 1]
```diff
@@ -0,0 +1,945 @@
...
+        $msg_cnt += $msgc;
+
+        warn("DEBUG: messages found in account: $acct : $msgc\n")
+          if ( $DEBUG > 0 );
+
+        if ($msgc) {
+            warn("------------------------\n");
+            foreach my $d (@$msgs) {
+
+                #my $tstr = scalar localtime( $d->{d} );
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: nit: remove this comment (old code I guess)

File: src/libexec/zmsearchndelete [hunk 2]
```diff
@@ -0,0 +1,945 @@
...
+                my $date = _date( $d->{d} );
+
+          #In testing, for very recently sent/received messages 'from' field
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: nit: fix comment indentation to match code indentation (comment should start on the same 'column' as the code)
nit: also add a space after the '#' for all the comments in this part of the code (above and below)
example:
`   #some word => # some word`

File: src/libexec/zmsearchndelete [hunk 3]
```diff
@@ -0,0 +1,945 @@
...
+          #is not getting populated instantaneously. It gets populated in
+          #next try or if 1st try is delayed by some time (?min) after mail sent
+                $d->{from} = "NA" unless ( defined( $d->{from} ) );
+
+                #my $kbsize = $d->{s} / 1024;
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: nit: remove old/commented out code

File: src/libexec/zmsearchndelete [hunk 4]
```diff
@@ -0,0 +1,945 @@
...
+                warn(
+                    "$acct\t\t$d->{id}\t$d->{cid}\t$date\t$d->{from}\t$d->{s}\n"
+                );
+
+           #create message ids string for MsgActionRequest (op=delete) SOAP CALL
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: nit: indent comment to same column as code in this block

File: src/libexec/zmsearchndelete [hunk 5]
```diff
@@ -0,0 +1,945 @@
...
+     --authuser admin@test.loc --password password \
+     --query "hrishi" --exclude phil@test.loc <input file>
+
+=head1 DESCRIPTION
+
+This utility is used to search particular string in messages for the
+specified account(s).
+It can be used to delete these messages on request.
+It uses the ZCS Admin SOAP APIs
+
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: nit: remove extra blank line

File: src/libexec/zmsearchndelete [hunk 6]
```diff
@@ -0,0 +1,945 @@
...
+        $Opt{go}
+    );
+
+    warn("$Prog: info: Total messages found for given query: $cnt \n");
+
+    my $tEnd = Benchmark->new;
+
+    warn( "$Prog: info: finished ", _ts(), "\n" );
+    warn( "$Prog: info: took ", timestr( timediff( $tEnd, $tStart ) ), "\n" );
+
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: nit: remove blank line

File: src/libexec/zmsearchndelete [hunk 7]
```diff
@@ -0,0 +1,945 @@
...
+          #next try or if 1st try is delayed by some time (?min) after mail sent
+                $d->{from} = "NA" unless ( defined( $d->{from} ) );
+
+                #my $kbsize = $d->{s} / 1024;
+                warn(
+                    "$acct\t\t$d->{id}\t$d->{cid}\t$date\t$d->{from}\t$d->{s}\n"
+                );
+
+           #create message ids string for MsgActionRequest (op=delete) SOAP CALL
+                $msg_ids .= $d->{id} . "," if ($go);
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: nit: if you wanted to learn a little perl, this $msg_ids (scalar variable) really should become an array @msg_ids... then delmsgs could be updated to take an array of message ids as arguments and generate the appropriate soap request with the data.  You can ignore this idea though if you don't want to learn a little bit more about how this (and Perl) works.

File: src/libexec/zmsearchndelete (hunk 1 above)
Comment (chinmay15) [reply src/libexec/zmsearchndelete]: done

File: src/libexec/zmsearchndelete (hunk 2 above)
Comment (chinmay15) [reply src/libexec/zmsearchndelete]: done

File: src/libexec/zmsearchndelete (hunk 3 above)
Comment (chinmay15) [reply src/libexec/zmsearchndelete]: done

File: src/libexec/zmsearchndelete (hunk 4 above)
Comment (chinmay15) [reply src/libexec/zmsearchndelete]: done

File: src/libexec/zmsearchndelete (hunk 5 above)
Comment (chinmay15) [reply src/libexec/zmsearchndelete]: done

File: src/libexec/zmsearchndelete (hunk 6 above)
Comment (chinmay15) [reply src/libexec/zmsearchndelete]: done

File: src/libexec/zmsearchndelete (hunk 7 above)
Comment (chinmay15) [reply src/libexec/zmsearchndelete]: used array @msg_ids and updated the same in delmsgs.

**Comment by chinmay15** [comment]
Updated. Please review.

File: src/libexec/zmsearchndelete [hunk 8]
```diff
@@ -0,0 +1,941 @@
...
+                # In testing, for very recently sent/received messages 'from' field
+                # is not getting populated instantaneously. It gets populated in
+                # next try or if 1st try is delayed by some time (?min) after mail sent
+                $d->{from} = "NA" unless ( defined( $d->{from} ) );
+
+                warn(
+                    "$acct\t\t$d->{id}\t$d->{cid}\t$date\t$d->{from}\t$d->{s}\n"
+                );
+
+               # create message ids string for MsgActionRequest (op=delete) SOAP CALL
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: nit: comment should be one more column to the right to align properly

File: src/libexec/zmsearchndelete [hunk 9]
```diff
@@ -0,0 +1,941 @@
...
+
+                delmsgs( { soap => $soap, head => $dhdr }, $acct, @msg_ids );
+            }
+        }
+    }
+    warn("------------------------\n");
+    return $msg_cnt;
+}
+
+# delmsgs( { soap => $soap, head => $dhdr }, $acct, @msg_ids )
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: nit: comment says @msg_ids, and that is the name we used in reportndelete() but in the delmsgs below the 's' was dropped from the variable, pick one and stick with it for consistency

File: src/libexec/zmsearchndelete [hunk 10]
```diff
@@ -0,0 +1,941 @@
...
+   500: real 21m19.636s user 20m42.027s sys 0m0.395s
+   200: real  9m36.827s user  8m52.900s sys 0m0.717s
+   100: real  5m47.554s user  4m50.981s sys 0m0.769s
+    75: real  4m51.761s user  3m54.414s sys 0m0.799s
+    50: real  4m17.056s user  3m02.626s sys 0m1.246s
+    40: real  4m22.345s user  2m57.274s sys 0m1.530s
+
+=head1 BUGS/KNOWN ISSUES
+
+=over 3
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: nit: change this to '4' from '3' (an indentation of 3 is unusual in POD)

File: src/libexec/zmsearchndelete [hunk 11]
```diff
@@ -0,0 +1,941 @@
...
+
+                warn(
+                    "$acct\t\t$d->{id}\t$d->{cid}\t$date\t$d->{from}\t$d->{s}\n"
+                );
+
+               # create message ids string for MsgActionRequest (op=delete) SOAP CALL
+                push @msg_ids, $d->{id} if ($go);
+            }
+
+            #execute 'delete messages' if --go option is specified
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: nit: space between '#' and 'execute'

File: src/libexec/zmsearchndelete [hunk 12]
```diff
@@ -0,0 +1,941 @@
...
+    my ( $soap, $ainfo ) = @{$opt}{qw(soap ainfo)};
+    my ($noadminauth) = @{$opt}{qw(noadminauth)};
+    my $ahdr = $ainfo->{header};
+    my ( $dhdr, $msg_cnt );
+    warn(
+        "------------------------
+    1:Account\t2:MSG_ID\t3:CONV_ID\t4:DATE(mm/dd/yy)\t5:FROM\t6:SIZE(Bytes)\n"
+    );
+
+    #report searched message details
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: nit: space between '#' and 'report'

File: src/libexec/zmsearchndelete [hunk 13]
```diff
@@ -0,0 +1,941 @@
...
+                warn(
+                    "$acct\t\t$d->{id}\t$d->{cid}\t$date\t$d->{from}\t$d->{s}\n"
+                );
+
+               # create message ids string for MsgActionRequest (op=delete) SOAP CALL
+                push @msg_ids, $d->{id} if ($go);
+            }
+
+            #execute 'delete messages' if --go option is specified
+            if ($go) {
```
Comment (plobbes) [inline src/libexec/zmsearchndelete]: Should we check if we have any message ids in @msg_ids (e.g. @msg_ids is not empty)? There's no reason to even bother doing any additional work here if it is empty.

File: src/libexec/zmsearchndelete (hunk 8 above)
Comment (chinmay15) [reply src/libexec/zmsearchndelete]: done

File: src/libexec/zmsearchndelete (hunk 9 above)
Comment (chinmay15) [reply src/libexec/zmsearchndelete]: done

File: src/libexec/zmsearchndelete (hunk 10 above)
Comment (chinmay15) [reply src/libexec/zmsearchndelete]: done

File: src/libexec/zmsearchndelete (hunk 13 above)
Comment (chinmay15) [reply src/libexec/zmsearchndelete]: done

**Comment by chinmay15** [comment]
Updated. Please review. Hoping to get approval now.

**Pull request merged**